`pyfunction!()`no longer works for private functions in other modules

### Bug Description

I realize this is probably desired behavior as it's weird (and possible not sound) to try to use a function that's not marked as public across a module boundary.  But, I wanted to open this more to document the behavior change as the error message wasn't entire clear to me when I first hit (so feel free to close this). I was apparently very sloppy with public vs private for my `pyfunction` usage because I assume the macros were previously always generating public functions regardless of whether the original was public or private. In trying to upgrade to pyo3 0.16.0 I came across a number of compiler errors that my python functions defined in other modules could not be found (see backtrace section for error message) when the function was not marked as public. This was a change from earlier releases as I could access the `pyfunction` in another module via the `wrap_pyfunction!` macro.

### Steps to Reproduce

`lib.rs`:

```rust
mod magic;

use magic::*;

use pyo3::prelude::*;
use pyo3::Python;
use pyo3::wrap_pyfunction;

#[pymodule]
fn test_stuff(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_wrapped(wrap_pyfunction!(my_magic_function))?;
    Ok(())
}
```

`magic.rs`:

```rust
use pyo3::prelude::*;

#[pyfunction]
fn my_magic_function(input: u64) -> u64 {
    2 * input
}
```



### Backtrace

```shell
Compiling test_stuff v0.1.0 (/tmp/test_stuff)
error[E0433]: failed to resolve: use of undeclared crate or module `my_magic_function`
  --> src/lib.rs:11:36
   |
11 |     m.add_wrapped(wrap_pyfunction!(my_magic_function))?;
   |                                    ^^^^^^^^^^^^^^^^^ use of undeclared crate or module `my_magic_function`

warning: unused import: `magic::*`
 --> src/lib.rs:3:5
  |
3 | use magic::*;
  |     ^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

For more information about this error, try `rustc --explain E0433`.
```
```


### Your operating system and version

Linux

### Your Python version (`python --version`)

Python 3.10

### Your Rust version (`rustc --version`)

Rust 1.59.0

### Your PyO3 version

0.16.0

### How did you install python? Did you use a virtualenv?

pacman

### Additional Info

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -61,6 +61,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - internal `handle_panic` helper [#2074](https://github.com/PyO3/pyo3/pull/2074) [#2158](https://github.com/PyO3/pyo3/pull/2158)
   - `#[pyfunction]` and `#[pymethods]` argument extraction [#2075](https://github.com/PyO3/pyo3/pull/2075) [#2085](https://github.com/PyO3/pyo3/pull/2085)
   - `#[pyclass]` type object creation [#2076](https://github.com/PyO3/pyo3/pull/2076) [#2081](https://github.com/PyO3/pyo3/pull/2081) [#2157](https://github.com/PyO3/pyo3/pull/2157)
+- Respect Rust privacy rules for items wrapped with `wrap_pyfunction` and `wrap_pymodule`. [#2081](https://github.com/PyO3/pyo3/pull/2081)
 - Add modulo argument to `__ipow__` magic method. [#2083](https://github.com/PyO3/pyo3/pull/2083)
 - Fix FFI definition for `_PyCFunctionFast`. [#2126](https://github.com/PyO3/pyo3/pull/2126)
 - `PyDateTimeAPI` and `PyDateTime_TimeZone_UTC` are are now unsafe functions instead of statics. [#2126](https://github.com/PyO3/pyo3/pull/2126)

--- a/guide/src/migration.md
+++ b/guide/src/migration.md
@@ -105,9 +105,9 @@ Because there is no such distinction from Python, implementing these methods wil
 
 The PyO3 behavior in 0.16 has been changed to be closer to this Python behavior by default.
 
-### `wrap_pymodule!` now respects privacy correctly
+### `wrap_pymodule!` and `wrap_pyfunction!` now respect privacy correctly
 
-Prior to PyO3 0.16 the `wrap_pymodule!` macro could use modules declared in Rust modules which were not reachable.
+Prior to PyO3 0.16 the `wrap_pymodule!` and `wrap_pyfunction!` macros could use modules and functions whose defining `fn` was not reachable according Rust privacy rules.
 
 For example, the following code was legal before 0.16, but in 0.16 is rejected because the `wrap_pymodule!` macro cannot access the `private_submodule` function:
 

